src/iso9660.rs

```diff
@@ -0,0 +1,364 @@
+}
+
+/// Reads the volume descriptor at cursor and advances to the next one.
+fn get_next_volume_descriptor(f: &mut fs::File) -> Result<Option<VolumeDescriptor>> {
+    const TYPE_BOOT: u8 = 0;
+    const TYPE_PRIMARY: u8 = 1;
+    const TYPE_SUPPLEMENTARY: u8 = 2;
+    const TYPE_TERMINATOR: u8 = 255;
+
+    let mut buf = vec![0; ISO9660_SECTOR_SIZE];
+    f.read_exact(&mut buf)
+        .context("reading volume descriptor")?;
+    let buf = &mut Bytes::from(buf);
+
+    Ok(match buf.get_u8() {
+        TYPE_BOOT => Some(VolumeDescriptor::Boot(parse_boot_descriptor(buf)?)),
+        TYPE_PRIMARY => Some(VolumeDescriptor::Primary(parse_primary_descriptor(buf)?)),
+        TYPE_SUPPLEMENTARY => Some(VolumeDescriptor::Supplementary),
+        TYPE_TERMINATOR => None,
+        t => Some(VolumeDescriptor::Unknown { type_id: t }),
+    })
+}
+
+/// Parses boot descriptor at cursor.
+fn parse_boot_descriptor(buf: &mut Bytes) -> Result<BootVolumeDescriptor> {
+    verify_descriptor_header(buf).context("parsing boot descriptor")?;
+    Ok(BootVolumeDescriptor {
+        boot_system_id: parse_iso9660_string(buf, 32)?,
+        boot_id: parse_iso9660_string(buf, 32)?,
+    })
+}
+
+/// Parses primary descriptor at cursor.
+fn parse_primary_descriptor(buf: &mut Bytes) -> Result<PrimaryVolumeDescriptor> {
+    verify_descriptor_header(buf).context("parsing primary descriptor")?;
+    let system_id = parse_iso9660_string(eat(buf, 1), 32).context("parsing system id")?;
+    let volume_id = parse_iso9660_string(buf, 32).context("parsing volume id")?;
+    let root = match get_next_directory_record(eat(buf, 156 - 72))? {
+        Some(DirectoryRecord::Directory(d)) => d,
+        _ => bail!("failed to parse root directory record from primary descriptor"),
+    };
+    Ok(PrimaryVolumeDescriptor {
+        system_id,
+        volume_id,
+        root,
+    })
+}
+
+/// Verifies descriptor header at cursor.
+fn verify_descriptor_header(buf: &mut Bytes) -> Result<()> {
+    const VOLUME_DESCRIPTOR_ID: &[u8] = b"CD001";
+    const VOLUME_DESCRIPTOR_VERSION: u8 = 1;
+
+    let id = buf.copy_to_bytes(5);
+    if id != VOLUME_DESCRIPTOR_ID {
+        bail!("unknown descriptor ID: {:?}", id);
+    }
+
+    let version = buf.get_u8();
+    if version != VOLUME_DESCRIPTOR_VERSION {
+        bail!("unknown descriptor version: {}", version);
+    }
+
+    Ok(())
+}
+
+pub struct IsoFsIterator {
+    dir: Bytes,
+}
+
+impl Iterator for IsoFsIterator {
+    type Item = Result<DirectoryRecord>;
+    fn next(&mut self) -> Option<Self::Item> {
+        get_next_directory_record(&mut self.dir)
+            .context("reading next record")
+            .transpose()
+    }
+}
+
+/// Reads the directory record at cursor and advances to the next one.
+fn get_next_directory_record(buf: &mut Bytes) -> Result<Option<DirectoryRecord>> {
+    let len = buf.get_u8() as usize;
+    if len == 0 {
+        return Ok(None);
+    }
+
+    let address = eat(buf, 1).get_u32_le();
```

Comment:
A corrupt image that puts a partial entry at the end of the directory will cause one of these calls to panic.  Do we want to check for that case?

Reply:
Good catch, fixed!